File: src/Alexander/validation/partition.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <dirent.h>
#include <errno.h>
#include "qsort.c"
// External function declarations - these are defined in your qsort.c
extern void swap(int *a, int *b);
extern int partition(int arr[], int low, int high);
extern void quickSort(int arr[], int low, int high);

char* extract_value(const char *json, const char *key) {
    char key_pattern[128];
    // Handle both quoted and unquoted keys, and potential whitespace
    snprintf(key_pattern, sizeof(key_pattern), "\"%s\"", key);
    
    const char *key_pos = strstr(json, key_pattern);
    if (!key_pos) {
        // Try without quotes around the key
        key_pos = strstr(json, key);
        if (!key_pos) return NULL;
        key_pos += strlen(key);
    } else {
        key_pos += strlen(key_pattern);
    }
    
    // Skip whitespace and colon
    while (*key_pos && (isspace(*key_pos) || *key_pos == ':')) {
        key_pos++;
    }
    
    // Expect opening quote
    if (*key_pos != '"') return NULL;
    key_pos++; // Skip opening quote
    
    const char *val_start = key_pos;
    const char *val_end = strchr(val_start, '"');
    if (!val_end) return NULL;
    
    size_t len = val_end - val_start;
    char *result = malloc(len + 1);
    if (!result) return NULL;
    
    memcpy(result, val_start, len);
    result[len] = '\0';
    return result;
}

unsigned char* parse_bytes(const char *str, size_t *byte_count) {
    // Handle format: "b'\\x00\\x01...'"
    const char *ptr = str;
    
    if (strncmp(ptr, "b'", 2) == 0 && ptr[strlen(ptr)-1] == '\'') {
        ptr += 2;  // Skip "b'"
    }
    
    // Count how many \x sequences we have
    size_t count = 0;
    const char *temp = ptr;
    while (*temp && *temp != '\'') {
        if (*temp == '\\' && *(temp + 1) == 'x') {
            count++;
            temp += 4;  // Skip \xHH
        } else {
            temp++;
        }
    }
    
    *byte_count = count;
    unsigned char *bytes = malloc(*byte_count);
    if (!bytes) return NULL;
    
    size_t i = 0;
    while (*ptr && *ptr != '\'' && i < count) {
        if (*ptr == '\\' && *(ptr + 1) == 'x') {
            // Parse hex digits
            char hex_str[3] = {ptr[2], ptr[3], '\0'};
            char *endptr;
            bytes[i] = (unsigned char)strtoul(hex_str, &endptr, 16);
            
            if (*endptr != '\0') {
                fprintf(stderr, "Invalid hex digits: %s\n", hex_str);
                free(bytes);
                return NULL;
            }
            i++;
            ptr += 4;  // Skip \xHH
        } else {
            ptr++;
        }
    }
    
    return bytes;
}

int main() {
    const char *dir_path = "data_out_partition";
    DIR *dir = opendir(dir_path);
    if (!dir) {
        fprintf(stderr, "Error opening directory '%s': %s\n", dir_path, strerror(errno));
        return EXIT_FAILURE;
    }

    struct dirent *entry;
    while ((entry = readdir(dir)) != NULL) {
        if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0)
            continue;
            
        char filepath[1024];
        snprintf(filepath, sizeof(filepath), "%s/%s", dir_path, entry->d_name);
        
        FILE *f = fopen(filepath, "r");
        if (!f) {
            fprintf(stderr, "Error opening file '%s': %s\n", filepath, strerror(errno));
            continue;
        }

        fseek(f, 0, SEEK_END);
        long size = ftell(f);
        fseek(f, 0, SEEK_SET);
        
        char *content = malloc(size + 1);
        if (!content) {
            fclose(f);
            continue;
        }
        
        fread(content, 1, size, f);
        fclose(f);
        content[size] = '\0';
        
        char *arr_str = extract_value(content, "arr");
        char *low_str = extract_value(content, "low");
        char *high_str = extract_value(content, "high");
        
        if (!arr_str || !low_str || !high_str) {
            fprintf(stderr, "JSON extraction failed in %s\n", filepath);
            free(arr_str);
            free(low_str);
            free(high_str);
            free(content);
            continue;
        }
        
        size_t arr_bytes, low_bytes, high_bytes;
        unsigned char *arr_bytes_arr = parse_bytes(arr_str, &arr_bytes);
        unsigned char *low_bytes_arr = parse_bytes(low_str, &low_bytes);
        unsigned char *high_bytes_arr = parse_bytes(high_str, &high_bytes);
        

        
        free(arr_str);
        free(low_str);
        free(high_str);
        free(content);
        
        if (!arr_bytes_arr || !low_bytes_arr || !high_bytes_arr) {
            fprintf(stderr, "Byte parsing failed in %s\n", filepath);
            free(arr_bytes_arr);
            free(low_bytes_arr);
            free(high_bytes_arr);
            continue;
        }

        // Convert byte arrays to integers (little-endian)
        int low = 0, high = 0;
        for (size_t i = 0; i < low_bytes; i++)
            low |= (low_bytes_arr[i] << (i * 8));
        for (size_t i = 0; i < high_bytes; i++)
            high |= (high_bytes_arr[i] << (i * 8));
            
        // Convert byte array to integer array
        if (arr_bytes % sizeof(int) != 0) {
            fprintf(stderr, "Invalid array size: %zu bytes\n", arr_bytes);
            free(arr_bytes_arr);
            free(low_bytes_arr);
            free(high_bytes_arr);
            continue;
        }
        
        int arr_len = arr_bytes / sizeof(int);
        int *arr = (int *)arr_bytes_arr;
        
        printf("Input array: [");
        for (int i = 0; i < arr_len; i++) {
            printf("%d%s", arr[i], (i < arr_len - 1) ? ", " : "");
        }
        printf("]\n");
        
        printf("Sorting range: low=%d, high=%d\n", low, high);
        
        // Validate bounds
        if (low < 0 || high >= arr_len || low > high) {
            fprintf(stderr, "Invalid bounds: low=%d, high=%d, arr_len=%d\n", low, high, arr_len);
            free(arr_bytes_arr);
            free(low_bytes_arr);
            free(high_bytes_arr);
            continue;
        }
        
        // Call your partition function from qsort.c
        int pivot_index = partition(arr, low, high);
        
        printf("Pivot index: %d\n", pivot_index);
        printf("\n");
        
        free(arr_bytes_arr);
        free(low_bytes_arr);
        free(high_bytes_arr);
    }
    
    closedir(dir);
    return EXIT_SUCCESS;
}

File: src/Alexander/validation/partition.rs
use std::fs;

mod qsort;

fn extract_value(json: &str, key: &str) -> Option<String> {
    let key_pattern = format!("\"{}\"", key);
    
    let key_pos = if let Some(pos) = json.find(&key_pattern) {
        pos + key_pattern.len()
    } else if let Some(pos) = json.find(key) {
        pos + key.len()
    } else {
        return None;
    };
    
    let remaining = &json[key_pos..];
    
    // Skip whitespace and colon
    let mut chars = remaining.chars();
    let mut start_pos = 0;
    for ch in chars.by_ref() {
        if ch == ':' || ch.is_whitespace() {
            start_pos += ch.len_utf8();
        } else {
            break;
        }
    }
    
    let value_part = &remaining[start_pos..];
    
    // Expect opening quote
    if !value_part.starts_with('"') {
        return None;
    }
    
    let value_part = &value_part[1..]; // Skip opening quote
    
    if let Some(end_pos) = value_part.find('"') {
        Some(value_part[..end_pos].to_string())
    } else {
        None
    }
}

fn parse_bytes(s: &str) -> Option<Vec<u8>> {
    let mut ptr = s;
    
    // Handle format: "b'\\x00\\x01...'"
    if ptr.starts_with("b'") && ptr.ends_with('\'') {
        ptr = &ptr[2..ptr.len()-1]; // Remove "b'" and trailing "'"
    }
    
    let mut bytes = Vec::new();
    let mut chars = ptr.chars().peekable();
    
    while let Some(ch) = chars.next() {
        if ch == '\\' && chars.peek() == Some(&'x') {
            chars.next(); // consume 'x'
            
            // Get next two hex digits
            let hex1 = chars.next()?;
            let hex2 = chars.next()?;
            
            let hex_str = format!("{}{}", hex1, hex2);
            let byte_val = u8::from_str_radix(&hex_str, 16).ok()?;
            bytes.push(byte_val);
        }
    }
    
    Some(bytes)
}

fn bytes_to_i32_le(bytes: &[u8]) -> i32 {
    let mut result = 0i32;
    for (i, &byte) in bytes.iter().enumerate() {
        result |= (byte as i32) << (i * 8);
    }
    result
}

fn bytes_to_i32_array_le(bytes: &[u8]) -> Vec<i32> {
    bytes.chunks_exact(4)
        .map(|chunk| {
            i32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]])
        })
        .collect()
}

// Partition function from qsort.rs module
use qsort::partition;

fn main() -> std::io::Result<()> {
    let dir_path = "data_out_partition";
    let entries = fs::read_dir(dir_path)?;
    
    for entry in entries {
        let entry = entry?;
        let path = entry.path();
        
        if path.is_file() {
            let content = fs::read_to_string(&path)?;
            
            let arr_str = extract_value(&content, "arr");
            let low_str = extract_value(&content, "low");
            let high_str = extract_value(&content, "high");
            
            if let (Some(arr_str), Some(low_str), Some(high_str)) = (arr_str, low_str, high_str) {
                let arr_bytes = parse_bytes(&arr_str);
                let low_bytes = parse_bytes(&low_str);
                let high_bytes = parse_bytes(&high_str);
                
                if let (Some(arr_bytes), Some(low_bytes), Some(high_bytes)) = (arr_bytes, low_bytes, high_bytes) {
                    let low = bytes_to_i32_le(&low_bytes);
                    let high = bytes_to_i32_le(&high_bytes);
                    let mut arr = bytes_to_i32_array_le(&arr_bytes);
                    
                    print!("Input array: [");
                    for (i, &val) in arr.iter().enumerate() {
                        if i > 0 { print!(", "); }
                        print!("{}", val);
                    }
                    println!("]");
                    
                    println!("Sorting range: low={}, high={}", low, high);
                    
                    let pivot_index = partition(Some(&mut arr), low, high);
                    
                    println!("Pivot index: {}", pivot_index);
                    println!();
                }
            }
        }
    }
    
    Ok(())
}
